File: web/src/lib/fx.ts
import "server-only";

/**
 * FX rate resolver — USD/ILS and friends, with a daily ECB rate from
 * frankfurter.app + an in-memory cache.
 *
 * Why we need it: Meta returns ad-account spend in the account's native
 * currency (`account_currency` field on insights). Aiweon's Meta ad account
 * is denominated in USD; `business_knowledge.monthly_budget_ils` and every
 * pace calculation downstream is in ILS. Without conversion the dashboard
 * displays "₪484" when the real spend is ~₪1,800 — same numeric value
 * mislabeled, which silently breaks every pace/budget reading the agent does
 * for non-ILS accounts.
 *
 * Why frankfurter.app: no API key, no rate limit for our usage (one fetch
 * per day per process), uses ECB reference rates which are the most widely
 * accepted "official" rate. Fallback to a hardcoded conservative rate if the
 * API is unreachable — better than crashing the dashboard.
 *
 * Cache: module-level Map keyed by `YYYY-MM-DD`. The rate changes ~once per
 * banking day; intra-day churn isn't worth chasing. Cache survives across
 * requests within the same Node process.
 */

/** Conservative fallback used when the FX API is unreachable. */
const USD_TO_ILS_FALLBACK = 3.7;

const rateCache = new Map<string, number>();

function todayKey(): string {
  return new Date().toISOString().slice(0, 10);
}

interface FrankfurterResponse {
  amount: number;
  base: string;
  date: string;
  rates: Record<string, number>;
}

/**
 * Resolve 1 USD → ILS for today. Cached by calendar date. Returns the
 * fallback rate (with `source: "fallback"`) on any network/parse error so
 * the caller can still render and surface "FX unavailable" if it cares.
 */
export async function getUsdToIlsRate(): Promise<{
  rate: number;
  source: "live" | "cached" | "fallback";
  date: string;
}> {
  const key = todayKey();
  const cached = rateCache.get(key);
  if (cached !== undefined) {
    return { rate: cached, source: "cached", date: key };
  }
  try {
    const res = await fetch(
      "https://api.frankfurter.app/latest?from=USD&to=ILS",
      {
        cache: "no-store",
        signal: AbortSignal.timeout(2500),
      },
    );
    if (!res.ok) throw new Error(`frankfurter HTTP ${res.status}`);
    const body = (await res.json()) as FrankfurterResponse;
    const rate = body.rates?.ILS;
    if (typeof rate !== "number" || !Number.isFinite(rate) || rate <= 0) {
      throw new Error(`frankfurter returned invalid ILS rate: ${rate}`);
    }
    rateCache.set(key, rate);
    return { rate, source: "live", date: body.date ?? key };
  } catch (e) {
    console.warn(
      `[fx] USD/ILS fetch failed, using fallback ${USD_TO_ILS_FALLBACK}: ${
        e instanceof Error ? e.message : "unknown"
      }`,
    );
    return { rate: USD_TO_ILS_FALLBACK, source: "fallback", date: key };
  }
}

/**
 * Convert an amount to ILS. `from === "ILS"` is a no-op. Currencies other
 * than USD currently fall through to identity — extend this when a real
 * non-USD/non-ILS account shows up. Returns the conversion metadata so the
 * UI can surface "מומר מ-$X (שער 3.71)".
 */
export async function convertToIls(
  amount: number,
  from: string | null | undefined,
): Promise<{
  amount_ils: number;
  rate_used: number;
  source_currency: string;
  fx_source: "live" | "cached" | "fallback" | "none";
}> {
  const cur = (from ?? "ILS").toUpperCase();
  if (cur === "ILS") {
    return {
      amount_ils: amount,
      rate_used: 1,
      source_currency: "ILS",
      fx_source: "none",
    };
  }
  if (cur === "USD") {
    const { rate, source } = await getUsdToIlsRate();
    return {
      amount_ils: amount * rate,
      rate_used: rate,
      source_currency: "USD",
      fx_source: source,
    };
  }
  // Unknown currency — pass through unchanged. The display layer should
  // catch `source_currency !== "ILS" && rate_used === 1` and warn.
  console.warn(`[fx] unsupported currency ${cur}, passing through 1:1`);
  return {
    amount_ils: amount,
    rate_used: 1,
    source_currency: cur,
    fx_source: "none",
  };
}

File: campaigner/lib/fx.py
"""
FX resolver — USD/ILS (and friends) using ECB daily rates from frankfurter.app.

Mirrors `web/src/lib/fx.ts`. Both sides need the same conversion because:

- Meta returns ad-account spend in the account's native currency
  (`account_currency` on the insights row). Aiweon's account is USD.
- `businesses.monthly_budget_ils` and every pace/threshold downstream is in
  ILS. Without conversion the morning `budget_health` row says
  "spend_this_month: 484" while the operator's mental model is "₪", which
  silently breaks pace, projected_monthly_spend, status, and every guardrail
  that compares ILS spend to ILS budget.

Cache: module-level dict keyed by `YYYY-MM-DD`. The rate changes ~once per
banking day; intra-day refresh isn't worth the latency. Survives across calls
inside a single tool invocation; cold-starts fetch fresh.

Fallback: 3.7 USD/ILS — conservative, hardcoded — when the API is unreachable.
The agent run still completes; the outputs blob marks `fx.rate_source =
"fallback"` so the human knows the FX line is approximate.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from datetime import UTC, datetime
from typing import Literal

USD_TO_ILS_FALLBACK = 3.7
FX_TIMEOUT_S = 2.5

# {date_iso: rate}
_rate_cache: dict[str, float] = {}


def _today_key() -> str:
    return datetime.now(UTC).date().isoformat()


def get_usd_to_ils_rate() -> tuple[float, Literal["live", "cached", "fallback"], str]:
    """
    Return (rate, source, date_iso). `source`:
      live      → fetched from frankfurter.app this call
      cached    → reused from module cache (already fetched today)
      fallback  → hardcoded USD_TO_ILS_FALLBACK after a fetch failure
    """
    key = _today_key()
    cached = _rate_cache.get(key)
    if cached is not None:
        return cached, "cached", key
    try:
        req = urllib.request.Request(
            "https://api.frankfurter.app/latest?from=USD&to=ILS",
            headers={"User-Agent": "campaigner-agent/1.0"},
        )
        with urllib.request.urlopen(req, timeout=FX_TIMEOUT_S) as resp:
            body = json.loads(resp.read().decode("utf-8"))
        rate = body.get("rates", {}).get("ILS")
        if not isinstance(rate, (int, float)) or rate <= 0:
            raise ValueError(f"invalid ILS rate in response: {rate!r}")
        rate_f = float(rate)
        _rate_cache[key] = rate_f
        return rate_f, "live", body.get("date") or key
    except (urllib.error.URLError, ValueError, OSError, TimeoutError) as e:
        # Quiet warning — tools route operator-visible errors through their
        # _contract.emit_*. This is a soft degradation, not a failure.
        import sys

        print(
            f"[fx] USD/ILS fetch failed, using fallback {USD_TO_ILS_FALLBACK}: {e}",
            file=sys.stderr,
        )
        return USD_TO_ILS_FALLBACK, "fallback", key


FxSource = Literal["live", "cached", "fallback", "none"]


def convert_to_ils(
    amount: float,
    from_currency: str | None,
) -> tuple[float, float, str, FxSource]:
    """
    Convert `amount` to ILS. Returns (amount_ils, rate_used, source_currency,
    fx_source).

    - from_currency=None or "ILS" → identity, rate_used=1, fx_source="none"
    - from_currency="USD" → ECB rate via `get_usd_to_ils_rate`
    - anything else → pass-through 1:1 + stderr warning. Extend when a real
      EUR/GBP account shows up.
    """
    cur = (from_currency or "ILS").upper()
    if cur == "ILS":
        return amount, 1.0, "ILS", "none"
    if cur == "USD":
        rate, source, _ = get_usd_to_ils_rate()
        return amount * rate, rate, "USD", source
    import sys

    print(
        f"[fx] unsupported currency {cur!r}, passing through 1:1",
        file=sys.stderr,
    )
    return amount, 1.0, cur, "none"
